R/dentist_qc.R

```diff
@@ -70,11 +73,180 @@ dentist <- function(zScore, LDmat, nSample,
+          fillStartIdx <- c(fillStartIdx, fillEndIdx[i - 1] + 1)
+          fill_end_pointer <- pos[windowStartIdx[i]] + 0.75 * window_size
+          fill_end_idx <- which.max(pos[pos <= fill_end_pointer])
+          fillEndIdx <- c(fillEndIdx, fill_end_idx)
+        }
+      }
+      # for the last window, the start is the next variant of the previous fillEndIdx, and the end is the last variant
+      fillStartIdx <- c(fillStartIdx, fillEndIdx[length(fillEndIdx)] + 1)
+      fillEndIdx <- c(fillEndIdx, length(pos))
+    }
+  }
+  # combine window information into a data frame
+  window_index <- 1:length(windowStartIdx)
+  window_divided_res <- data.frame(
+    windowIdx = window_index,
+    windowStartIdx = windowStartIdx,
+    windowEndIdx = windowEndIdx,
+    fillStartIdx = fillStartIdx,
+    fillEndIdx = fillEndIdx
+  )
+  # check if the divided windows are valid
+  invalid_ends <- window_divided_res$fillStartIdx[1] != 1 | window_divided_res$fillEndIdx[nrow(window_divided_res)] != length(pos)
+  if (nrow(window_divided_res) == 1 & invalid_ends) {
+    stop("Invalid window divided!")
+  } else if (nrow(window_divided_res) > 1) {
+    invalid_middle <- 0
+    for (i in 1:(nrow(window_divided_res) - 1)) {
+      if (fillStartIdx[i + 1] - fillEndIdx[i] != 1) {
+        invalid_middle <- invalid_middle + 1
+      }
+    }
+    if (invalid_ends | invalid_middle != 0) {
+      stop("Invalid window divided!")
+    }
+  }
+  return(window_divided_res)
+}
+
+
+merge_windows <- function(imputed_result_by_window, window_divided_res) {
+  if (length(imputed_result_by_window) != nrow(window_divided_res)) {
+    stop("Different number of windows and imputed results!")
+  }
+  merged_results <- c()
+  for (k in 1:nrow(window_divided_res)) {
+    imputed_k <- imputed_result_by_window[[k]]
+    imputed_k$index_within_window <- seq(1:nrow(imputed_k))
+    imputed_k <- imputed_k %>%
+      mutate(index_global = index_within_window + window_divided_res$windowStartIdx[k] - 1)
+    extracted_results <- imputed_k %>%
+      filter(index_global >= window_divided_res$fillStartIdx[k] & index_global <= window_divided_res$fillEndIdx[k])
+    merged_results <- rbind(merged_results, extracted_results)
+  }
+  return(merged_results)
+}
+
+dentist_detect_outliers <- function(sum_stat, LDmat, nSample,
+                                    window_size = 2000000, pValueThreshold = 5e-8, propSVD = 0.4, gcControl = FALSE,
+                                    nIter = 10, gPvalueThreshold = 0.05, ncpus = 1, seed = 999, correct_chen_et_al_bug = TRUE) {
+  # detect for column names and order by pos
+  if (!any(tolower(c("pos", "position")) %in% tolower(colnames(sum_stat))) ||
+    !any(tolower(c("z", "zscore")) %in% tolower(colnames(sum_stat)))) {
+    stop("Input sum_stat is missing either 'pos'/'position' or 'z'/'zscore' column.")
+  }
+  sum_stat <- sum_stat %>% arrange(pos)
+  ### FIXME: formalize this logic somewhere
```

Comment:
You can use AI to help you integrate this to the roxygen2 documentation string. You don't have that yet anyways. Just copy your entire function to one of the AI chats asking it to convert the FIXME comment, along with the interface of the function, into roxygen2 documentation. And you add it here.